website/addons/googledrive/tests/test_models.py

```diff
@@ -7,313 +7,366 @@
+        assert_is_none(self.node_settings.drive_folder_id)
+        assert_is_none(self.node_settings.user_settings)
+        assert_is_none(self.node_settings.folder_path)
+        assert_is_none(self.node_settings.drive_folder_name)
 
-        assert_in('node', params)
-        assert_in('folder', params)
-        assert_in('project', params)
-        assert_equal(last_log.action, 'googledrive_node_deauthorized')
+    def test_set_target_folder(self):
 
-    def test_set_folder(self):
-        folder_name = {
-            'id': '1234',
-            'name': 'freddie',
-            'path': 'queen/freddie',
+        folder = {
+            'id': 'fake-folder-id',
+            'name': 'fake-folder-name',
+            'path': 'fake_path'
         }
 
-        self.node_settings.set_folder(folder_name, auth=Auth(self.user))
+
+        external_account = ExternalAccountFactory()
+        self.user.external_accounts.append(external_account)
+        self.user.save()
+
+        self.node_settings.set_auth(
+            external_account=external_account,
+            user=self.user,
+            )
+
+        assert_is_none(self.node_settings.drive_folder_id)
+
+        self.node_settings.set_target_folder(
+            folder,
+            auth=Auth(user=self.user),
+            )
+
+        # instance was updated
+        assert_equal(
+            self.node_settings.drive_folder_id,
+            'fake-folder-id',
+            )
+
+        has_access = self.user_settings.verify_oauth_access(
+            node=self.node,
+            external_account=external_account,
+            metadata={'folder': 'fake-folder-id'}
+        )
+
+        # user_settings was updated
+        assert_true(has_access)
+
+        log = self.node.logs[-1]
+        assert_equal(log.action, 'googledrive_folder_selected')
+        assert_equal(log.params['folder_id'], folder['id'])
+        assert_equal(log.params['folder_name'], folder['path'])
+
+    def test_has_auth_false(self):
+        external_account = ExternalAccountFactory()
+
+        assert_false(self.node_settings.has_auth)
+
+        # both external_account and user_settings must be set to have auth
+        self.node_settings.external_account = external_account
+        assert_false(self.node_settings.has_auth)
+
+        self.node_settings.external_account = None
+        self.node_settings.user_settings = self.user_settings
+        assert_false(self.node_settings.has_auth)
+
+        # set_auth must be called to have auth
+        self.node_settings.external_account = external_account
+        self.node_settings.user_settings = self.user_settings
+        assert_false(self.node_settings.has_auth)
+
+    def test_has_auth_true(self):
+        external_account = ExternalAccountFactory()
+        self.user.external_accounts.append(external_account)
+
+        self.node_settings.set_auth(external_account, self.user)
+
+        # mendeley_list_id should have no effect
```

Comment:
>         # mendeley_list_id should have no effect 

Given that this is the Google drive addon, I concur. Worth glancing at this section in case there are any other copy/paste issues.